File: sdk/device/device.test.ts
import { describe, expect, it } from 'vitest'

import {
  DeviceCapabilityGrantState,
  DeviceCapabilityLocalState,
  DeviceCapabilityState,
  DeviceCheckoutRootAccess,
  DeviceLiveness,
  DeviceSetupState,
  type Device,
} from './device.pb.js'
import {
  DeviceCapabilityKindFilesystem,
  DeviceCapabilityKindForgeWorker,
  deviceCheckoutRootCanRead,
  deviceCheckoutRootCanWrite,
  findDeviceCapabilityByKind,
  findReadableDeviceCheckoutRoot,
  findSelectableDeviceCheckoutRoot,
  findSelectableDeviceForgeWorker,
  findWritableDeviceCheckoutRoot,
  hasSelectableDeviceCapabilityKind,
  isDeviceCapabilitySelectable,
  isDeviceSelectable,
} from './device.js'

describe('Device selection helpers', () => {
  it('accepts lima as a workflow-selectable Device target', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      lastStatus: {
        liveness: DeviceLiveness.ONLINE,
        message: 'device session ready',
      },
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/skiffos-checkout',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: 'skiffos',
            displayPath: '~/repos/skiffos/skiffos',
            selectionRef: 'device/lima/filesystem/skiffos',
            access: DeviceCheckoutRootAccess.READ_WRITE,
            readAvailable: true,
            writeAvailable: true,
          },
          policy: {
            localPolicyRef: 'device/lima/filesystem/skiffos',
            grantPolicyRef: 'space/grants/lima/skiffos',
            localState: DeviceCapabilityLocalState.ENABLED,
            grantState: DeviceCapabilityGrantState.ALLOWED,
          },
        },
        {
          id: 'forge-worker',
          kind: DeviceCapabilityKindForgeWorker,
          label: 'Forge Worker',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'forge/workers/lima',
            typeId: 'forge/worker',
          },
          policy: {
            localPolicyRef: 'device/lima/forge-worker',
            grantPolicyRef: 'space/grants/lima/forge-worker',
          },
        },
      ],
    }

    expect(isDeviceSelectable(device)).toBe(true)
    expect(
      hasSelectableDeviceCapabilityKind(device, DeviceCapabilityKindFilesystem),
    ).toBe(true)
    expect(
      hasSelectableDeviceCapabilityKind(
        device,
        DeviceCapabilityKindForgeWorker,
      ),
    ).toBe(true)
    expect(
      findDeviceCapabilityByKind(device, DeviceCapabilityKindFilesystem)?.link
        ?.objectKey,
    ).toBe('unixfs/skiffos-checkout')
    const checkoutRoot = findSelectableDeviceCheckoutRoot(
      device,
      'skiffos',
    )?.checkoutRoot
    expect(checkoutRoot?.selectionRef).toBe('device/lima/filesystem/skiffos')
    expect(deviceCheckoutRootCanRead(checkoutRoot)).toBe(true)
    expect(deviceCheckoutRootCanWrite(checkoutRoot)).toBe(true)
    expect(
      findReadableDeviceCheckoutRoot(device, 'skiffos')?.link?.typeId,
    ).toBe('unixfs/fs-node')
    expect(
      findWritableDeviceCheckoutRoot(device, 'skiffos')?.link?.objectKey,
    ).toBe('unixfs/skiffos-checkout')
    expect(findSelectableDeviceForgeWorker(device)?.link?.objectKey).toBe(
      'forge/workers/lima',
    )
  })

  it('does not select blocked or disabled capability summaries', () => {
    expect(
      isDeviceCapabilitySelectable({
        id: 'checkout-root-skiffos',
        kind: DeviceCapabilityKindFilesystem,
        label: 'SkiffOS checkout',
        state: DeviceCapabilityState.GRANT_BLOCKED,
      }),
    ).toBe(false)
    expect(
      isDeviceCapabilitySelectable({
        id: 'forge-worker',
        kind: DeviceCapabilityKindForgeWorker,
        label: 'Forge Worker',
        state: DeviceCapabilityState.DISABLED,
      }),
    ).toBe(false)
  })

  it('requires a filesystem owner link before a checkout root is readable', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          checkoutRoot: {
            name: 'skiffos',
            access: DeviceCheckoutRootAccess.READ_ONLY,
            readAvailable: true,
          },
        },
      ],
    }

    expect(findSelectableDeviceCheckoutRoot(device, 'skiffos')).toBeTruthy()
    expect(findReadableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
  })

  it('rejects unavailable filesystem and forge-worker capabilities', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.DISABLED,
          link: {
            objectKey: 'unixfs/skiffos-checkout',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: 'skiffos',
            access: DeviceCheckoutRootAccess.READ_ONLY,
            readAvailable: true,
          },
        },
        {
          id: 'forge-worker',
          kind: DeviceCapabilityKindForgeWorker,
          label: 'Forge Worker',
          state: DeviceCapabilityState.DECLARED,
          link: {
            objectKey: 'forge/workers/lima',
            typeId: 'forge/worker',
          },
        },
      ],
    }

    expect(findReadableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
    expect(findSelectableDeviceForgeWorker(device)).toBeUndefined()
  })

  it('skips blocked same-kind capabilities when checking availability', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-blocked',
          kind: DeviceCapabilityKindFilesystem,
          label: 'Blocked checkout',
          state: DeviceCapabilityState.DISABLED,
        },
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
        },
      ],
    }

    expect(
      hasSelectableDeviceCapabilityKind(device, DeviceCapabilityKindFilesystem),
    ).toBe(true)
  })

  it('requires explicit checkout-root access before treating availability as readable', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/skiffos-checkout',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: 'skiffos',
            readAvailable: true,
          },
        },
      ],
    }

    const checkoutRoot = findSelectableDeviceCheckoutRoot(
      device,
      'skiffos',
    )?.checkoutRoot
    expect(deviceCheckoutRootCanRead(checkoutRoot)).toBe(false)
    expect(findReadableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
  })

  it('requires allowed policy before treating a checkout root as writable', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: DeviceCapabilityKindFilesystem,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/skiffos-checkout',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: 'skiffos',
            access: DeviceCheckoutRootAccess.READ_WRITE,
            readAvailable: true,
            writeAvailable: true,
          },
          policy: {
            localState: DeviceCapabilityLocalState.ENABLED,
            grantState: DeviceCapabilityGrantState.BLOCKED,
          },
        },
      ],
    }

    expect(findReadableDeviceCheckoutRoot(device, 'skiffos')).toBeTruthy()
    expect(findWritableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
  })

  it('skips blocked same-kind forge workers when checking availability', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'forge-worker-blocked',
          kind: DeviceCapabilityKindForgeWorker,
          label: 'Blocked Forge Worker',
          state: DeviceCapabilityState.DISABLED,
          link: {
            objectKey: 'forge/workers/blocked',
            typeId: 'forge/worker',
          },
        },
        {
          id: 'forge-worker',
          kind: DeviceCapabilityKindForgeWorker,
          label: 'Forge Worker',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'forge/workers/lima',
            typeId: 'forge/worker',
          },
        },
      ],
    }

    expect(findSelectableDeviceForgeWorker(device)?.link?.objectKey).toBe(
      'forge/workers/lima',
    )
  })

  it('does not throw on partially populated legacy capability records', () => {
    const device: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'missing-kind',
          label: 'Missing kind',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/missing-kind',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            readAvailable: true,
            access: DeviceCheckoutRootAccess.READ_ONLY,
          },
        },
        {
          id: 'missing-name',
          kind: DeviceCapabilityKindFilesystem,
          label: 'Missing name',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/missing-name',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            access: DeviceCheckoutRootAccess.READ_ONLY,
            readAvailable: true,
          },
        },
      ],
    }

    expect(
      hasSelectableDeviceCapabilityKind(device, DeviceCapabilityKindFilesystem),
    ).toBe(true)
    expect(findSelectableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
    expect(findReadableDeviceCheckoutRoot(device, 'skiffos')).toBeUndefined()
    expect(findSelectableDeviceForgeWorker(device)).toBeUndefined()
  })

  it('trims identity and owner links before treating Device state as selectable', () => {
    expect(
      isDeviceSelectable({
        peerId: '   ',
        label: 'lima',
        setupState: DeviceSetupState.DEVICE_SESSION_READY,
      }),
    ).toBe(false)

    const whitespaceOwner: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: ` ${DeviceCapabilityKindFilesystem} `,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: '   ',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: ' skiffos ',
            access: DeviceCheckoutRootAccess.READ_ONLY,
            readAvailable: true,
          },
        },
        {
          id: 'forge-worker',
          kind: DeviceCapabilityKindForgeWorker,
          label: 'Forge Worker',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'forge/workers/lima',
            typeId: '   ',
          },
        },
      ],
    }

    expect(
      findReadableDeviceCheckoutRoot(whitespaceOwner, 'skiffos'),
    ).toBeUndefined()
    expect(findSelectableDeviceForgeWorker(whitespaceOwner)).toBeUndefined()

    const trimmed: Device = {
      peerId: '12D3KooWLima',
      label: 'lima',
      setupState: DeviceSetupState.DEVICE_SESSION_READY,
      capabilities: [
        {
          id: 'checkout-root-skiffos',
          kind: ` ${DeviceCapabilityKindFilesystem} `,
          label: 'SkiffOS checkout',
          state: DeviceCapabilityState.AVAILABLE,
          link: {
            objectKey: 'unixfs/skiffos-checkout',
            typeId: 'unixfs/fs-node',
          },
          checkoutRoot: {
            name: ' skiffos ',
            access: DeviceCheckoutRootAccess.READ_ONLY,
            readAvailable: true,
          },
        },
      ],
    }
    expect(
      findReadableDeviceCheckoutRoot(trimmed, ' skiffos ')?.link?.objectKey,
    ).toBe('unixfs/skiffos-checkout')
  })
})

File: sdk/device/device_test.go
package s4wave_device

import "testing"

func TestDeviceValidatePinsIdentityAndCapabilityIds(t *testing.T) {
	dev := &Device{
		PeerId: "12D3KooWDevice",
		Label:  "Build Host",
		Capabilities: []*DeviceCapability{
			{Id: "filesystem", Kind: "filesystem", Label: "Files"},
			{Id: "forge-worker", Kind: "forge-worker", Label: "Forge Worker"},
		},
	}
	if err := dev.Validate(); err != nil {
		t.Fatalf("valid device failed validation: %v", err)
	}

	dev.PeerId = ""
	if err := dev.Validate(); err == nil {
		t.Fatal("expected missing peer_id to fail validation")
	}

	dev.PeerId = "12D3KooWDevice"
	dev.Capabilities = append(dev.Capabilities, &DeviceCapability{Id: "filesystem"})
	if err := dev.Validate(); err == nil {
		t.Fatal("expected duplicate capability id to fail validation")
	}
}

func TestDeviceValidateCapabilityLinkRequiresType(t *testing.T) {
	err := (&Device{
		PeerId: "peer-device",
		Label:  "Device",
		Capabilities: []*DeviceCapability{{
			Id:   "filesystem",
			Kind: "filesystem",
			Link: &DeviceCapabilityLink{ObjectKey: "files/root"},
		}},
	}).Validate()
	if err == nil {
		t.Fatal("Validate accepted capability object link without type_id")
	}
}

func TestDeviceSelectableLimaWorkflowTarget(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		LastStatus: &DeviceStatus{
			Liveness: DeviceLiveness_DEVICE_LIVENESS_ONLINE,
			Message:  "device session ready",
		},
		Capabilities: []*DeviceCapability{
			{
				Id:    "checkout-root-skiffos",
				Kind:  DeviceCapabilityKindFilesystem,
				Label: "SkiffOS checkout",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
				Link: &DeviceCapabilityLink{
					ObjectKey: "unixfs/skiffos-checkout",
					TypeId:    "unixfs/fs-node",
				},
				CheckoutRoot: &DeviceCheckoutRootCapability{
					Name:           "skiffos",
					DisplayPath:    "~/repos/skiffos/skiffos",
					SelectionRef:   "device/lima/filesystem/skiffos",
					Access:         DeviceCheckoutRootAccess_DEVICE_CHECKOUT_ROOT_ACCESS_READ_WRITE,
					ReadAvailable:  true,
					WriteAvailable: true,
				},
				Policy: &DeviceCapabilityPolicy{
					LocalPolicyRef: "device/lima/filesystem/skiffos",
					GrantPolicyRef: "space/grants/lima/skiffos",
					LocalState:     DeviceCapabilityLocalState_DEVICE_CAPABILITY_LOCAL_STATE_ENABLED,
					GrantState:     DeviceCapabilityGrantState_DEVICE_CAPABILITY_GRANT_STATE_ALLOWED,
				},
			},
			{
				Id:    "forge-worker",
				Kind:  DeviceCapabilityKindForgeWorker,
				Label: "Forge Worker",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
				Link: &DeviceCapabilityLink{
					ObjectKey: "forge/workers/lima",
					TypeId:    "forge/worker",
				},
				Policy: &DeviceCapabilityPolicy{
					LocalPolicyRef: "device/lima/forge-worker",
					GrantPolicyRef: "space/grants/lima/forge-worker",
					LocalState:     DeviceCapabilityLocalState_DEVICE_CAPABILITY_LOCAL_STATE_ENABLED,
					GrantState:     DeviceCapabilityGrantState_DEVICE_CAPABILITY_GRANT_STATE_ALLOWED,
				},
			},
		},
	}
	if err := dev.Validate(); err != nil {
		t.Fatalf("lima workflow target failed validation: %v", err)
	}
	if !dev.IsSelectable() {
		t.Fatal("expected device to be selectable")
	}
	if !dev.HasSelectableCapabilityKind(DeviceCapabilityKindFilesystem) {
		t.Fatal("expected selectable filesystem capability")
	}
	if !dev.HasSelectableCapabilityKind(DeviceCapabilityKindForgeWorker) {
		t.Fatal("expected selectable forge-worker capability")
	}
	if got := dev.FindSelectableForgeWorker().GetLink().GetObjectKey(); got != "forge/workers/lima" {
		t.Fatalf("expected forge worker object link, got %q", got)
	}
	if got := dev.FindCapabilityByKind(DeviceCapabilityKindFilesystem).GetLink().GetObjectKey(); got != "unixfs/skiffos-checkout" {
		t.Fatalf("expected checkout root object link, got %q", got)
	}
	if got := dev.FindSelectableCheckoutRoot("skiffos").GetCheckoutRoot().GetSelectionRef(); got != "device/lima/filesystem/skiffos" {
		t.Fatalf("expected checkout root selection ref, got %q", got)
	}
	readable := dev.FindReadableCheckoutRoot("skiffos")
	if readable == nil {
		t.Fatal("expected readable checkout root")
	}
	if readable.GetLink().GetTypeId() != "unixfs/fs-node" {
		t.Fatalf("expected unixfs owner link, got %q", readable.GetLink().GetTypeId())
	}
	if !DeviceCheckoutRootCanRead(readable.GetCheckoutRoot()) {
		t.Fatal("expected checkout root to be readable")
	}
	if !DeviceCheckoutRootCanWrite(readable.GetCheckoutRoot()) {
		t.Fatal("expected checkout root to be writable after approval")
	}
	if writable := dev.FindWritableCheckoutRoot("skiffos"); writable == nil {
		t.Fatal("expected checkout root to be write-gateable")
	}
}

func TestDeviceCapabilitySelectionRejectsBlockedStates(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{{
			Id:     "checkout-root-skiffos",
			Kind:   DeviceCapabilityKindFilesystem,
			Label:  "SkiffOS checkout",
			State:  DeviceCapabilityState_DEVICE_CAPABILITY_STATE_GRANT_BLOCKED,
			Detail: "blocked by Space grant",
			Link: &DeviceCapabilityLink{
				ObjectKey: "unixfs/skiffos-checkout",
				TypeId:    "unixfs/fs-node",
			},
		}},
	}
	if !dev.IsSelectable() {
		t.Fatal("expected device identity and setup state to remain selectable")
	}
	if dev.HasSelectableCapabilityKind(DeviceCapabilityKindFilesystem) {
		t.Fatal("grant-blocked capability was selectable")
	}
	if dev.FindSelectableCheckoutRoot("skiffos") != nil {
		t.Fatal("grant-blocked checkout root was selectable")
	}
	if dev.FindReadableCheckoutRoot("skiffos") != nil {
		t.Fatal("grant-blocked checkout root was readable")
	}
}

func TestDeviceValidateCheckoutRootShape(t *testing.T) {
	err := (&Device{
		PeerId: "peer-device",
		Label:  "Device",
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindTerminal,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:          "skiffos",
				ReadAvailable: true,
			},
		}},
	}).Validate()
	if err == nil {
		t.Fatal("Validate accepted checkout_root on non-filesystem capability")
	}

	err = (&Device{
		PeerId: "peer-device",
		Label:  "Device",
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:           "skiffos",
				WriteAvailable: true,
			},
		}},
	}).Validate()
	if err == nil {
		t.Fatal("Validate accepted checkout_root write availability without read availability")
	}

	err = (&Device{
		PeerId: "peer-device",
		Label:  "Device",
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:          "skiffos",
				ReadAvailable: true,
			},
		}},
	}).Validate()
	if err == nil {
		t.Fatal("Validate accepted checkout_root availability without access mode")
	}

	err = (&Device{
		PeerId: "peer-device",
		Label:  "Device",
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:           "skiffos",
				Access:         DeviceCheckoutRootAccess_DEVICE_CHECKOUT_ROOT_ACCESS_READ_ONLY,
				ReadAvailable:  true,
				WriteAvailable: true,
			},
		}},
	}).Validate()
	if err == nil {
		t.Fatal("Validate accepted read-only checkout_root access with write availability")
	}
}

func TestDeviceCheckoutRootAccessGatesAvailability(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			Link: &DeviceCapabilityLink{
				ObjectKey: "unixfs/skiffos-checkout",
				TypeId:    "unixfs/fs-node",
			},
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:          "skiffos",
				ReadAvailable: true,
			},
		}},
	}
	if DeviceCheckoutRootCanRead(dev.GetCapabilities()[0].GetCheckoutRoot()) {
		t.Fatal("unknown checkout_root access was readable")
	}
	if dev.FindReadableCheckoutRoot("skiffos") != nil {
		t.Fatal("unknown checkout_root access produced readable owner link")
	}
}

func TestDeviceReadableCheckoutRootRequiresOwnerLink(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:          "skiffos",
				Access:        DeviceCheckoutRootAccess_DEVICE_CHECKOUT_ROOT_ACCESS_READ_ONLY,
				ReadAvailable: true,
			},
		}},
	}
	if dev.FindSelectableCheckoutRoot("skiffos") == nil {
		t.Fatal("expected checkout root to remain selectable")
	}
	if dev.FindReadableCheckoutRoot("skiffos") != nil {
		t.Fatal("checkout root without owner link was readable")
	}
}

func TestDeviceCapabilityPolicyRejectsUnavailablePocInputs(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{
			{
				Id:    "checkout-root-skiffos",
				Kind:  DeviceCapabilityKindFilesystem,
				Label: "SkiffOS checkout",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_DISABLED,
				Link: &DeviceCapabilityLink{
					ObjectKey: "unixfs/skiffos-checkout",
					TypeId:    "unixfs/fs-node",
				},
				CheckoutRoot: &DeviceCheckoutRootCapability{
					Name:          "skiffos",
					Access:        DeviceCheckoutRootAccess_DEVICE_CHECKOUT_ROOT_ACCESS_READ_ONLY,
					ReadAvailable: true,
				},
				Policy: &DeviceCapabilityPolicy{
					LocalState: DeviceCapabilityLocalState_DEVICE_CAPABILITY_LOCAL_STATE_DISABLED,
					GrantState: DeviceCapabilityGrantState_DEVICE_CAPABILITY_GRANT_STATE_ALLOWED,
				},
			},
			{
				Id:    "forge-worker",
				Kind:  DeviceCapabilityKindForgeWorker,
				Label: "Forge Worker",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_DECLARED,
				Link: &DeviceCapabilityLink{
					ObjectKey: "forge/workers/lima",
					TypeId:    "forge/worker",
				},
				Policy: &DeviceCapabilityPolicy{
					LocalState: DeviceCapabilityLocalState_DEVICE_CAPABILITY_LOCAL_STATE_ENABLED,
					GrantState: DeviceCapabilityGrantState_DEVICE_CAPABILITY_GRANT_STATE_UNKNOWN,
				},
			},
		},
	}
	if dev.FindReadableCheckoutRoot("skiffos") != nil {
		t.Fatal("disabled checkout root was readable")
	}
	if dev.FindWritableCheckoutRoot("skiffos") != nil {
		t.Fatal("disabled checkout root was writable")
	}
	if dev.FindSelectableForgeWorker() != nil {
		t.Fatal("declared forge worker was selectable")
	}
}

func TestDeviceWritableCheckoutRootRequiresAllowedPolicy(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{{
			Id:    "checkout-root-skiffos",
			Kind:  DeviceCapabilityKindFilesystem,
			Label: "SkiffOS checkout",
			State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			Link: &DeviceCapabilityLink{
				ObjectKey: "unixfs/skiffos-checkout",
				TypeId:    "unixfs/fs-node",
			},
			CheckoutRoot: &DeviceCheckoutRootCapability{
				Name:           "skiffos",
				Access:         DeviceCheckoutRootAccess_DEVICE_CHECKOUT_ROOT_ACCESS_READ_WRITE,
				ReadAvailable:  true,
				WriteAvailable: true,
			},
			Policy: &DeviceCapabilityPolicy{
				LocalState: DeviceCapabilityLocalState_DEVICE_CAPABILITY_LOCAL_STATE_ENABLED,
				GrantState: DeviceCapabilityGrantState_DEVICE_CAPABILITY_GRANT_STATE_BLOCKED,
			},
		}},
	}
	if dev.FindReadableCheckoutRoot("skiffos") == nil {
		t.Fatal("expected blocked write policy to keep checkout root readable")
	}
	if dev.FindWritableCheckoutRoot("skiffos") != nil {
		t.Fatal("grant-blocked checkout root was writable")
	}
}

func TestDeviceCapabilitySelectionSkipsBlockedSameKind(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{
			{
				Id:    "checkout-root-blocked",
				Kind:  DeviceCapabilityKindFilesystem,
				Label: "Blocked checkout",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_DISABLED,
			},
			{
				Id:    "checkout-root-skiffos",
				Kind:  DeviceCapabilityKindFilesystem,
				Label: "SkiffOS checkout",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
			},
		},
	}
	if !dev.HasSelectableCapabilityKind(DeviceCapabilityKindFilesystem) {
		t.Fatal("expected later selectable filesystem capability")
	}
}

func TestDeviceFindSelectableCapabilityByID(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWDevice",
		Label:      "Build Host",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{
			{
				Id:    "glados",
				Kind:  "sensor",
				Label: "Fixture Capability",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
				Link:  &DeviceCapabilityLink{ProtocolId: "plugin/test-plugin/fixture/v0"},
			},
			{
				Id:    "blocked-sensor",
				Kind:  "sensor",
				Label: "Blocked Capability",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_GRANT_BLOCKED,
				Link:  &DeviceCapabilityLink{ProtocolId: "plugin/test-plugin/blocked/v0"},
			},
		},
	}

	got := dev.FindSelectableCapability("glados")
	if got == nil || got.GetId() != "glados" {
		t.Fatalf("expected glados capability, got %v", got)
	}
	if dev.FindSelectableCapability(" blocked-sensor ") != nil {
		t.Fatal("blocked capability must not be selectable by id")
	}
	if dev.FindSelectableCapability("") != nil {
		t.Fatal("empty id must not select any capability")
	}
	if dev.FindSelectableCapability("plugin/test-plugin/fixture/v0") != nil {
		t.Fatal("raw protocol id must not select a capability")
	}
	if (*Device)(nil).FindSelectableCapability("glados") != nil {
		t.Fatal("nil device must not select a capability")
	}
}

func TestDeviceForgeWorkerSelectionSkipsBlockedSameKind(t *testing.T) {
	dev := &Device{
		PeerId:     "12D3KooWLima",
		Label:      "lima",
		SetupState: DeviceSetupState_DEVICE_SETUP_STATE_DEVICE_SESSION_READY,
		Capabilities: []*DeviceCapability{
			{
				Id:    "forge-worker-blocked",
				Kind:  DeviceCapabilityKindForgeWorker,
				Label: "Blocked Forge Worker",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_DISABLED,
				Link: &DeviceCapabilityLink{
					ObjectKey: "forge/workers/blocked",
					TypeId:    "forge/worker",
				},
			},
			{
				Id:    "forge-worker",
				Kind:  DeviceCapabilityKindForgeWorker,
				Label: "Forge Worker",
				State: DeviceCapabilityState_DEVICE_CAPABILITY_STATE_AVAILABLE,
				Link: &DeviceCapabilityLink{
					ObjectKey: "forge/workers/lima",
					TypeId:    "forge/worker",
				},
			},
		},
	}
	if got := dev.FindSelectableForgeWorker().GetLink().GetObjectKey(); got != "forge/workers/lima" {
		t.Fatalf("expected later selectable forge worker link, got %q", got)
	}
}
